Support experimental features

bahmutov opened the pull request:
- Closes #6257


### User facing changelog

<!--
Explain the change(s) for every user to read in our changelog.
-->

### Additional details

- boolean flags only for now from `process.env.CYPRESS_EXPERIMENTS`. Unknown experiments are ignored
- trying it in action for component testing https://github.com/cypress-io/cypress/pull/5923
- on the Node side use `packages/server/lib/experiments.ts`

```js
// example
const experiments = require('./experiments').getExperiments()
if (experiments.componentTesting) {
 ...
}
```
- on the desktop GUI side, we grab the environment variable `EXPERIMENTS` from resolved config and parse it using the same logic from `packages/server`. Which means technically the experiment can be overwritten using plugins file or CLI for the browser and desktop GUI, while only getting value from the environment variable on the Node side

### How has the user experience changed?

There is a new tab and by default, since there is nothing whitelisted in the config, it is just notification text

<img width="1248" alt="Screen Shot 2020-02-07 at 4 56 28 PM" src="https://user-images.githubusercontent.com/2212006/74069447-3a63da80-49cc-11ea-9829-3475c77e637f.png">


### PR Tasks

<!-- 
These tasks must be completed before a PR is merged.
Delete tasks if they are not applicable. 
-->

- [ ] have GUI and terminal changes been approved?

Then:
- [x] Have tests been added/updated?
- [ ] Has the original issue been tagged with a release in ZenHub? <!-- (internal team only)-->
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [ ] On links updated and released for https://on.cypress.io/experiments

Plan is to merge this PR with `componentTesting` or empty list of experiments (?) and then merge https://github.com/cypress-io/cypress/pull/5923 with `componentTesting` experiment (and a few tests for new mounting)

bahmutov committed:
add packages/server experiments

bahmutov committed:
show experiments in desktop GUI

bahmutov commented:
Added Desktop GUI tab showing experiments

<img width="1247" alt="Screen Shot 2020-01-28 at 12 04 06 PM" src="https://user-images.githubusercontent.com/2212006/73287044-98bbdc80-41c6-11ea-9f1e-8eed4eea10d6.png">

bahmutov committed:
add intro experiments text

bahmutov committed:
conditional message

bahmutov commented:
## UI review

Desktop GUI experiments tab

"Learn more" link goes to (non-existent yet) https://on.cypress.io/experiments

### Experiment enabled

<img width="1232" alt="Screen Shot 2020-01-28 at 1 54 12 PM" src="https://user-images.githubusercontent.com/2212006/73295703-2e129d00-41d6-11ea-816e-86e44a455f41.png">

### No experiments enabled

<img width="1212" alt="Screen Shot 2020-01-28 at 1 55 04 PM" src="https://user-images.githubusercontent.com/2212006/73295721-38cd3200-41d6-11ea-880f-0691d8bd5749.png">

- [ ] looks good, @brian-mann ?

bahmutov committed:
add experiments record to start of the headless run

bahmutov commented:
## UI Review

Headless run with experiments enabled (without experiments same terminal output)

<img width="1144" alt="Screen Shot 2020-01-28 at 2 19 45 PM" src="https://user-images.githubusercontent.com/2212006/73297780-2f45c900-41da-11ea-9ff6-97996680af87.png">

- [ ] looks good, @brian-mann ?

flotwig commented on packages/desktop-gui/src/settings/experiments.jsx:
```diff
@@ -0,0 +1,58 @@
+      </div>
+
+      <div>
+        <h3><input type="checkbox" readOnly defaultChecked={experiments.componentTesting} /> component testing</h3>
```
i don't know how crazy I am about using a disabled checkbox to distinguish between enabled and disabled - would prefer "Status: Enabled" or "Status: Disabled" instead, for example

bahmutov replied:
wait, something is weird, this is missing some of my commits!

bahmutov replied:
now it is good!

flotwig commented on packages/desktop-gui/src/settings/experiments.jsx:
```diff
@@ -0,0 +1,69 @@
+      </div>
+
+      <div>
+        <h3>{experiments.componentTesting ? '✔️' : '🚫'} component testing</h3>
```
oh yeah, this is what i was talking about. i feel like the emoji are unclear, can we use "status: enabled/disabled" or something instead?

bahmutov replied:
look at the screenshots above, they are beautiful https://github.com/cypress-io/cypress/pull/6265#issuecomment-579402112

bahmutov replied:
but I am not against something like `status: enabled` or `status: disabled` labels instead.

flotwig replied:
yeah, I think that would be clearer

jennifer-shehane commented:
I'm gonna work on the Desktop-GUI to freshen the design up for the experiements flags display.

bahmutov commented:
## Feedback - changes requested

- instead of creating new wrapper and only accepting environment variable, put experiments into the existing config. That way the flags can be passed in `cypress.json`, CLI, environment variable or overwritten by the plugins. When you change the `cypress.json`, the test runner will reload, just like now, and would not require every user setting an environment variable.
- name experimental properties in the config starting with `experimental_`. This way the Experiments settings tab can still show them nicely

brian-mann commented:
@bahmutov I would camel case the experiments...

`experimentalComponentTesting`
vs
`experimental_componentTesting`

bahmutov commented:
Updated logic to just use `config`, still have to clean up code a little bit from previous code before pushing.

Experimental properties are just keys in the config

<img width="587" alt="Screen Shot 2020-01-29 at 3 14 20 PM" src="https://user-images.githubusercontent.com/2212006/73394032-cda15f80-42aa-11ea-90ce-9ddeac26c923.png">

The experiments tab in Desktop GUI picks those properties and shows them nicely if they are non-default

<img width="1212" alt="Screen Shot 2020-01-29 at 3 29 28 PM" src="https://user-images.githubusercontent.com/2212006/73394741-29201d00-42ac-11ea-822c-6bde4186786d.png">

bahmutov committed:
keep experiments in the config, remove duplicate logic

bahmutov committed:
small tweaks

bahmutov committed:
remove experimentalComponentTesting for now, separate PR

jennifer-shehane commented:
When we talked about this earlier, I assumed this would be a nested object, which seems cleaner to me than a bunch of very long config values starting with `experiemental`, but....I guess this complicates things when wanting to just turn it on with one env var.  

```
"experiments": {
  "componentTesting": true,
}
```

Looked at some other tools cause I wasn't totally satisfied.

- **Docker** seems to have you set `experimental: "enabled"`, and then just list the features you want out separately, which I don't think is great either. https://docs.docker.com/engine/reference/commandline/cli/#experimental-features
- **Node** does have everything prefixed with `experimental` like `--experimental-json-modules` https://nodejs.org/api/cli.html
- **Chrome** also has an `experimental` flag outright to turn them all on or off and then separate flags which are all prefixed `experimental`. https://developer.chrome.com/extensions/experimental
- **Kubernetes** uses `feature-gates` 😆 https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/ Although they do have a nice feature where you can say, in the CLI print all of the experimental features available (probably overkill for us know though, they have 100's of features).
- Bazel has `experimental` prefix https://docs.bazel.build/versions/2.0.0/command-line-reference.html
- Parcel `experiemental` prefix.
- Everything else I saw doesn't explicitly specify 'experiemental' as part of actually setting the value and solely has it mentioned in documentation.


I think what looks weird is the camelCasing. Nobody camelCases the values but I get wanting to stick with our camelCase convention.

jennifer-shehane commented:
Clicking into the Experiments panel completely borks if you do not have any experiments defined. I've updated the title to be `[WIP]` as this is not ready for review.

jennifer-shehane committed:
rearrange settings spec into sections per panel

jennifer-shehane commented:
My first iteration. I decided to pull inspiration from the Dashboard, tried to get a similar design for their Beta features - the main difference being they call theirs 'Cypress Labs' and ours is called 'Experiments'.

**Current Design**

<img width="800" alt="Screen Shot 2020-01-30 at 2 49 47 PM" src="https://user-images.githubusercontent.com/1271364/73431881-cc863600-436f-11ea-9acc-6bb1f17deac6.png">


**Dashboard**

<img width="960" alt="Screen Shot 2020-01-30 at 2 48 46 PM" src="https://user-images.githubusercontent.com/1271364/73431826-abbde080-436f-11ea-8eab-ea8055e89d3c.png">

jennifer-shehane commented:
I started to write test in the Desktop Gui specs for the Experiments panel, but not nearly enough. I also updated the code to iterate through 'experiments' and not blow up when there are no experiment variables defined. 

This code is a bit naive. We won't be able to pull any key starting with `experimental`, we will have to explicitly whitelist all keys because people could just make up keys and it would do nothing. https://github.com/cypress-io/cypress/blob/experiments-6257/packages/server/lib/experiments.ts#L35:L35

We also need to build up more properties for each experiment for the display:

- title/name
- value
- default value
- description

bahmutov commented:
@jennifer-shehane did you push any code? I don't see any commits. I love the UI with toggle - but these are read-only, right? since the user has to change them using `cypress.json` or CLI or env for now?

We still control the experiments, thus the user cannot just put _any_ `experimentalX` config value, we still whitelist keys, and only take the ones we know about.

As far as props for experiments: 'title/name', 'value', 'default value', 'description' - we have `value` coming from config object (resolved), title and description are the text in `experiments.jsx` and the only missing thing I think is `default value`. Unfortunately, since the experiment is part of config, our config has to have `resolved` objects with current `value` and `defaultValue` and `from` fields. We can tackle this in a separate PR - the config coffeescript code is already hairy enough I feel, I would 

- land this PR after more tweaks and tests
- rewrite config in TypeScript (at least the interface)
- add `default value` property to each resolved config value

bahmutov committed:
do not crash if there are no experiments

bahmutov committed:
return component testing experiment

jennifer-shehane committed:
work on styling of experiments panel + adding basic tests

jennifer-shehane committed:
Merge remote-tracking branch 'origin/experiments-6257' into experiments-6257

# Conflicts:
#	packages/desktop-gui/src/settings/experiments.jsx

bahmutov committed:
update settings spec text

bahmutov committed:
add unit tests for experimental features in cypress run output

jennifer-shehane commented:
It is read only. They are not toggles, just display in the Desktop-GUI, although Brian suggested that they could be toggles (we just write the value directly to their `cypress.json`).

bahmutov committed:
Merge branch 'develop' into experiments-6257

jennifer-shehane committed:
Add styling for 'non-enabled' status sign.

jennifer-shehane reviewed (changes requested):
Sorry, maybe I was not clear in my explanation before. 

1. This line will `.map` through each experiment found:

https://github.com/cypress-io/cypress/blob/experiments-6257/packages/desktop-gui/src/settings/experiments.jsx#L28:L28

Right now the test for 'Component Testing' and 'Enables the use of ....' is statically written into the template. There needs to be a title and a description for each experimental feature so that when it loops through each future experiement this data can be shown. 

So an `experiment` should be an object constructed like the one below before it reaches the `.jsx` template:

```js
{
  title: "Component Testing"
  description: "Enables the use of component testing using framework-specific libraries to mount your component directly from the spec file."
  value: true
  default: false
}
```

2. This code is naive. It grabs any configuration value starting with `experimental`, which means if I pass the below, then `experimentalFooBar` will be shown in the Desktop Gui (but likely just crash the Desktop-GUI Settings panel in this case - and/or show itself under 'Component Testing' section due to #1 issue described above). 

We have to whitelist the experiment keys available. 

```js
{
  "chromeWebSecurity": false,
  "experimentalFooBar": true
}
```

3. There are inadequate tests in the Desktop Gui testing the display of the Component Testing toggle and other experimental keys being read in correctly. I added a single test just to get them started, but this is not testing that any of the content or the toggle is displaying correctly at all. https://github.com/cypress-io/cypress/pull/6265/files#diff-bb3b13ed2f43f3f1f9486a5593115e69R614

4. ~The `OFF` toggle shows as green.~ Fixed in https://github.com/cypress-io/cypress/commit/b5d1d87dd52bb23af93e5d7d3450504efcb8aeee

I'm not sure why `WIP` was removed from the title when neither of my previous comments from before about these things were addressed.

bahmutov commented:
So, it is hard to work on this branch if the real experiment is in https://github.com/cypress-io/cypress/pull/5923 and I have updated that branch based on your UI (and removed `.map`, we need to be explicit)

<img width="943" alt="Screen Shot 2020-02-03 at 8 25 13 AM" src="https://user-images.githubusercontent.com/2212006/73656886-0f415a00-465f-11ea-953e-e30941575fb9.png">

I also added Desktop GUI tests there, because it should set the experiment status. 

Ughh, this is what having two long-living pull requests means, neither is working very well. But to address your comments again:

- we are whitelisting config keys, thus unknown `experimental*` keys will be skipped anyway. We have to whitelist, because I suspect each `experimental*` key will also need some other `cypress.json` config keys. For example, component testing needs component tests folder setting.

So to move this forward, and land this pull request to actually do work for experimental component testing (that's the only thing that matters), I will keep your `.map` here, will move a few tests from https://github.com/cypress-io/cypress/pull/5923/commits/105b9c85ed9e6f97cb080e7be905cd9d160a7121 into this branch so at least we will know if desktop gui works. The real work will be in components testing PR

bahmutov committed:
add more experiments and key whitelist in experiments GUI

bahmutov commented:
I have added more tests, added whitelist to ensure unknown experimental features do not show component testing experiment

<img width="1357" alt="Screen Shot 2020-02-03 at 9 01 13 AM" src="https://user-images.githubusercontent.com/2212006/73659274-e8d1ed80-4663-11ea-9d6e-a4e2a19b4066.png">

bahmutov committed:
Merge branch 'develop' into experiments-6257

jennifer-shehane commented:
@bahmutov Yeah, I'm not sure why you need the other repo at all. I never ran it with this PR, just mock everything in the tests.

jennifer-shehane reviewed (dismissed):
@bahmutov Yeah, I'm not sure why you need the other repo at all. I never ran it with this PR, you can mock everything in the tests. 

You mentioned 'removing map', but maybe I'm not understanding what you are meaning. 

In this code in this PR, we are looping over the experiments in the JSX template and showing the exact same title and description for every single experiment that would ever be passed to it. 

https://github.com/cypress-io/cypress/blob/experiments-6257/packages/desktop-gui/src/settings/experiments.jsx#L33:L33

It is my understanding that this PR (NOT the component testing PR) is meant to support **any** experimental features that we deem necessary in the future. 

Right now the React template is static and will literally print 'Component Testing' and the description for component testing for any experimental feature added in the future. So I don't see this PR as meeting the basic function of supporting 'Experimental Features'.

bahmutov committed:
Merge branch 'develop' into experiments-6257

bahmutov commented:
There is nothing to be afraid - there is a whitelist of experiments right now, and only component testing is whitelisted (and this is in addition to whitelisting experimental keys in the config object)

```
// whilelist known experiments here. Note that "experimental*" keys are also
// whitelisted in the config code
const isKnownExperiment = (experiment, key) =>
  ['experimentalComponentTesting'].includes(key)
```

and this PR already has e2e tests with unknown and known experiments, https://github.com/cypress-io/cypress/pull/6265#issuecomment-581427919

In other words, nothing bad can happen, we should be good to go, it will be blank (just the experiments preamble text) once merged into develop.

bahmutov committed:
Merge branch 'develop' into experiments-6257

bahmutov committed:
add message when there are no experiments

bahmutov committed:
Merge branch 'develop' into experiments-6257

bahmutov committed:
Merge branch 'develop' into experiments-6257

bahmutov committed:
Merge branch 'develop' into experiments-6257

bahmutov committed:
Merge branch 'develop' into experiments-6257

bahmutov committed:
Merge branch 'develop' into experiments-6257

bahmutov committed:
fix coffeelint

bahmutov committed:
fix merge in settings spec

bahmutov committed:
trailing whitespace

bahmutov committed:
Merge branch 'develop' into experiments-6257

bahmutov committed:
Merge branch 'experiments-6257' of github.com:cypress-io/cypress into experiments-6257

bahmutov committed:
Merge branch 'develop' into experiments-6257

bahmutov committed:
make experiments info all live in one place

bahmutov committed:
add ts lint exception

bahmutov commented:
@jennifer-shehane I have reworked the PR to be more to your liking - all experimental summaries and names are in a single place in server/lib/experiments.ts file and experiments.jsx is all dynamic and nothing is hardcoded.

bahmutov committed:
update unit test

bahmutov committed:
update run spec

bahmutov committed:
do not show experiments panel at all without experiments

bahmutov committed:
add more comments

bahmutov commented:
- I have hidden the "Experiments" panel if there are no experiments
- the complex config objects - I think it makes sense to look at this in the components testing PR, since it will be a concrete question there

bahmutov committed:
Merge branch 'develop' into experiments-6257

jennifer-shehane committed:
Update style of ON/OFF status for experimental features.

jennifer-shehane committed:
Merge branch 'develop' into experiments-6257

jennifer-shehane commented:
Updated the styles of the ON/OFF switch, so new design up for feedback.

<img width="798" alt="Screen Shot 2020-03-13 at 3 48 19 PM" src="https://user-images.githubusercontent.com/1271364/76607402-5239fb80-6542-11ea-8529-840cab2b3a2c.png">

<img width="800" alt="Screen Shot 2020-03-13 at 3 49 02 PM" src="https://user-images.githubusercontent.com/1271364/76607435-59f9a000-6542-11ea-85ba-423fdf93a23d.png">

jennifer-shehane committed:
remove empty messaging from experiments panel (since panel is no longer shown when empty)

jennifer-shehane committed:
remove check for 'isEmpty' when showing list

jennifer-shehane reviewed (commented):
Overall this PR looks good to me, barring approval for the design changes on the 'Experiments' panel

bahmutov commented on packages/desktop-gui/src/settings/experiments.jsx:
```diff
@@ -22,30 +22,28 @@ const Experiments = observer(({ project }) => {
           If you'd like to try out what we're working on, you can enable these beta features for your project by setting configuration using the <code>experimental</code> prefix.
         </p>
       </div>
-      {!_.isEmpty(experiments) &&
```
good catch

bahmutov committed:
Merge branch 'develop' into experiments-6257

bahmutov committed:
update lock file
